articles/avoid-fix-flaky-tests.md

```diff
@@ -0,0 +1,73 @@
+---
+authorName: Adrian Herzog
+authorGithubUsername: adiherzog
+issue: 42
+title: How to avoid / deal with flaky system tests
+---
+# {{page.title}}
+
+There are many reasons to avoid automated tests that check an entire system or even a group of connected systems. But there are also good reasons to build and maintain a reasonable number of such tests. For this article, let's assume that the tests are there for a good reason and provide valuable information to the team.
+
+Now, what can you do if one or more of these tests are flaky? By flaky I mean that the test changes from green to red or vice versa without any changes to the code.
+
+Here are a number of strategies that hopefully will help you keeping the tests stable.
+
+
+# Increase Your Chances To Succeed
+
+## Only Write Automated Tests For What's Under Your Control
+
+If the system under tests (SUT) containts components which are not under your control, it's very hard to keep the SUT in a stable state. For those systems not under your control, you might not know whether they have changed in the meantime or not. And even if you do know, you are most probably not able to influence them and change them back to how you need them.
+
+Therefore don't rely on other teams systems if possible. Mock them away so you have full control. Create System Tests instead of System Integration Tests. And of course these system tests must be written by the people who also program the system itself.
+
+If there's no way around System Integration Tests, try to get control over other systems as much as possible. E.g. align very closely with these systems regarding test data and deployments.
+
+## Avoid Broken Windows
+
+As soon as one test fails, insist on fixing it. Otherwise people might start ignoring failures and you'll end up in a [broken window](https://en.wikipedia.org/wiki/Broken_windows_theory) situation. There's a root cause to every flaky test. Either it's the test itself that needs improvement or it's a bug in the software.
+
+## Get The Timing Right
+
+Most issues with flaky tests are related to timing. Know your testing framework well, e.g. when it waits for something and when it doesn't. If you're using Selenium or Protractor, learn about [explicit and implicit waits](https://www.seleniumhq.org/docs/04_webdriver_advanced.jsp) and use them wisely (I strongly recommend using explicit waits). If your system under test has asynchronous behaviour, make sure you understand it well and write the tests accordingly. Configure sensible timeouts.
```

Comment:
Testing time sensitive systems is an art (as I have painfully discovered in the last year).
Explicit timings are a trap - it is much better to have clear break conditions and acceptable timeouts (a la soft real time) than mess around.
That is if I understand the phrasing correctly, because my advice is "control time": Never use system or computer time in tests, make sure you explicitly set references (whether dates or intervals). 
The best case scenario is if you can artificially define "slices" or "ticks" whose increment is controlled by your test harness. 
Asynchronous systems' behaviour varies wildly depending on load, execution platform, test assumptions and test level.
Handling time is also very different depending on the class of tests you have: Web GUI is most susceptible and also the most difficult to control, desktop GUI follows but can be structured for controlled easier etc.
Also, I would try to phrase it avoiding mention of tools and products, since the inherent problems in the tool design (Selenium in particular) usually invalidate the best practice we would like to follow.

Reply:
I removed the reference to Selenium and Protractor and rephrased the advice. Please have another look once I have committed it. Maybe you have an additional point to add to the list.